src/OpenApi/src/Extensions/JsonNodeSchemaExtensions.cs

```diff
@@ -327,10 +327,19 @@ internal static void ApplyParameterInfo(this JsonNode schema, ApiParameterDescri
     /// <param name="createSchemaReferenceId">A delegate that generates the reference ID to create for a type.</param>
     internal static void ApplyPolymorphismOptions(this JsonNode schema, JsonSchemaExporterContext context, Func<JsonTypeInfo, string?> createSchemaReferenceId)
     {
-        // The `context.Path.Length == 0` check is used to ensure that we only apply the polymorphism options
+        // The `context.BaseTypeInfo == null` check is used to ensure that we only apply the polymorphism options
         // to the top-level schema and not to any nested schemas that are generated.
-        if (context.TypeInfo.PolymorphismOptions is { } polymorphismOptions && context.Path.Length == 0)
+        if (context.TypeInfo.PolymorphismOptions is { } polymorphismOptions && context.BaseTypeInfo == null)
         {
+            // System.Text.Json supports serializing to a non-abstract base class if no discriminator is provided.
+            // OpenAPI requires that all polymorphic sub-schemas have an associated discriminator. To maintain accuracy,
+            // with the actual implementation and comply with the limitations of the OpenAPI schema we do not emit
+            // a `discriminator` property for types that are represented this way.
+            if (!context.TypeInfo.Type.IsAbstract &&
+                !polymorphismOptions.DerivedTypes.Any(type => type.DerivedType == context.TypeInfo.Type))
```

Comment:
I'm guessing `polymorphismOptions` includes an exhaustive list of base-type/subtype pairs?

Reply:
It's exhaustive in the sense that it matches what the serialier uses to determine derived types for polymorphic schemas. System.Text.Json populates this list based on the [JsonDerivedType attributes](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/polymorphism?pivots=dotnet-8-0#serialize-properties-of-derived-classes) on the base type.

Reply:
Okay, I was pretty confused by the name of the `DerivedType` property.  In this case `type` is a `JsonDerivedType`, which is effectively a type info/wrapper and `type.DerivedType` is the wrapped type.  That is, `DerivedType` is the type wrapped by the `JsonDerivedType` and not a type derived from it.

Reply:
(I had originally thought each element of `DerivedTypes` probably contained both `BaseType` and `DerivedType` properties and was based on the type system, rather than attributes on the base type).